File: P187_Dijkstras_algorithm.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution{
    public:
	// Using Set
    // Time complexity O(ElogV) || Space complexity O(E+V)
    vector <int> dijkstra(int V, vector<vector<int>> adj[], int S)
    {
        set<pair<int, int>> st;
        vector<int> distance(V, 1e9);
        
        st.insert({0, S});
        distance[S] = 0;
        
        while(!st.empty()){
            auto it = *(st.begin());
            int node = it.second;
            int dis = it.first;
            st.erase(it);
            
            for(auto it : adj[node]){
                int adjNode = it[0];
                int edgWt = it[1];
                
                if(dis + edgWt < distance[adjNode]){
                    // Erase if it existed
                    if(distance[adjNode] != 1e9){
                        st.erase({distance[adjNode], adjNode});
                    }
                    
                    distance[adjNode] = dis + edgWt;
                    st.insert({distance[adjNode], adjNode});
                }
            }
        }
        
        return distance;
    }
    };

File: P187_Dijkstras_algorithm.java
import java.util.*;

class Pair{
    int node, dist;
    
    Pair(int a, int b){
        node = a;
        dist = b;
    }
}

class P187_Dijkstras_algorithm{

    // Using Priority Queue
    // Time complexity O(ElogV) || Space complexity O(E+V)
    static int[] dijkstra(int V, ArrayList<ArrayList<ArrayList<Integer>>> adj, int S)
    {
        int distance[] = new int[V];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[S] = 0;
        
        PriorityQueue<Pair> pq = new PriorityQueue<>((x,y) -> x.dist - y.dist);
        // <node, distance>
        pq.add(new Pair(S, distance[S]));
        
        while(!pq.isEmpty()){
            
            Pair current = pq.poll();
            int currNode = current.node;
            int currDist = current.dist;
            
            for(ArrayList<Integer> nbr : adj.get(currNode)){
                int nbrNode = nbr.get(0);
                int nbrDistance = nbr.get(1);
                
                if(currDist + nbrDistance < distance[nbrNode]){
                    distance[nbrNode] = currDist + nbrDistance;
                    pq.add(new Pair(nbrNode, distance[nbrNode]));
                }
            }
        }
        
        return distance;
    }

    public static void main(Strint args[]){

    }
}